Fix: Corrected file paths when generating CTs, APIs, Controllers, Middlewares, Policies, Services inside Plugins

sunnysonx opened the pull request:
### What does it do?

Fixed file paths for Generator when creating them inside the plugin:
- CTs
- APIs
- Controllers
- Middlewares
- Policies
- Services

### Why is it needed?

Currently, when you generate a plugin through CLI, all `CTs, APIs, controllers, middlewares, policies, services` files are located inside `server` directory: `src/plugins/{pluginName}/server/`,  and that's the correct behavior. 
![image](https://user-images.githubusercontent.com/14160283/147662120-357921e9-91dc-4c77-b8d1-07d4aaa290e9.png)

But when you try to generate separately one of the following `CTs, APIs, controllers, middlewares, policies, services` inside a plugin, these are generated directly in the `src/plugins/{pluginName}` directory, instead of having them generated in /server folder. So you end up having two directories. In that case all the  `CTs, APIs, controllers, middlewares, policies, services` that are generated directly inside plugin's directly do not have any effect and are not imported to strapi-server.js
![image](https://user-images.githubusercontent.com/14160283/147662495-791ec8f5-ddc9-4bb5-965e-c010987fd479.png)

Elhebert commented on packages/generators/generators/lib/plops/content-type.js:
```diff
@@ -80,7 +81,7 @@ module.exports = plop => {
       }, {});
 
-      const filePath = getFilePath(answers.destination);
+      const filePath = resolveFilePath(answers.plugin,getFilePath(answers.destination));
```
You can, instead, update the `getFilePath` method to add the `/server` suffix to the plugin path:

https://github.com/strapi/strapi/blob/3706033d9fbf994376f31567d3adb659779854fd/packages/generators/generators/lib/plops/utils/get-file-path.js#L9

sunnysonx replied:
done

Elhebert commented on packages/generators/generators/lib/plops/controller.js:
```diff
@@ -18,7 +18,8 @@ module.exports = plop => {
     ],
     actions(answers) {
-      const filePath = getFilePath(answers.destination);
+      const resolveFilePath = (isPlugin,filePath) => isPlugin ? `${filePath}/server` : filePath;
+      const filePath = resolveFilePath(answers.plugin,getFilePath(answers.destination));
```
Same as for the CT generator. This should be fixed inside the `getFilePath` method.

sunnysonx replied:
done

Elhebert commented on packages/generators/generators/lib/plops/service.js:
```diff
@@ -16,7 +16,8 @@ module.exports = plop => {
     ],
     actions(answers) {
-      const filePath = getFilePath(answers.destination);
+      const resolveFilePath = (isPlugin,filePath) => isPlugin ? `${filePath}/server` : filePath;
+      const filePath = resolveFilePath(answers.plugin,getFilePath(answers.destination));
```
Same as for the CT generator. This should be fixed inside the `getFilePath` method.

sunnysonx replied:
done

Elhebert reviewed (changes requested):
Thanks for the PR 👌

I would suggest updated the path in the `getFilePath` utils instead of adding a new conditional to the generators.

sunnysonx commented:
@Elhebert 
-Removed/reverted the `/server` path that I have added in previous commits in generators
-Updated the path in the getFilePath utils and added `/server` there
-Refactored middlewares/policies and added the same logic with getFilePath utils (these contained a different logic because these can be created inside root folder)
-Refactored cli prompt answers in get-destinations-prompts to allow 'root' value for middlewares/policies.

Please review

Elhebert commented on packages/generators/generators/lib/plops/policy.js:
```diff
@@ -15,14 +16,7 @@ module.exports = plop => {
-        filePath = `./`;
-      }
+      const filePath = getFilePath(answers.destination);
 
```
You can't use `getFilePath` in this case.

The "default" value returned by `getFilePath` is `api/{{ id }}`, where as the policy default value is `./`.
Either you adapt the `getFilePath` method to have an argument to set a default value or you leave the code the way it was.

Both options are valid. I think that keeping the code as is might be the simplest as we don't add complexity to the `getFilePath` function for 2 use cases.

Elhebert commented on packages/generators/generators/lib/plops/middleware.js:
```diff
@@ -15,15 +16,7 @@ module.exports = plop => {
-        filePath = `./`;
-      }
-
+      const filePath = getFilePath(answers.destination);
```
Same as for the `policy` generator.

Elhebert commented on packages/generators/generators/lib/plops/api.js:
```diff
@@ -45,7 +46,8 @@ module.exports = plop => {
     actions(answers) {
-      const filePath = answers.isPluginApi && answers.plugin ? 'plugins/{{plugin}}/server' : 'api/{{id}}';
+      const answerDestination = answers.isPluginApi && answers.plugin ? 'plugin' : 'api';
+      const filePath = getFilePath(answerDestination);
```
`getFilePath` isn't compatible with this use case.

The API generator `filePath` wil either be `plugins/ {{ plugin }}/server` or `api/{{ id }}`. But with you changes it will be `plugins/ {{ plugin }}/server` or `api/{{ api }}`.

I would revert this change for simplicity and avoiding adding more code than necessary.

Elhebert reviewed (changes requested):
It's a step in the right direction. However you also updated generators to use `getFilePath` when it's not compatible with them.

I added some comments on how to fix them.

Elhebert commented:
You can disregard my last review. I don't know why GitHub wasn't showing all the changes 🤔

Elhebert reviewed (changes requested):
LGTM. One comment about the API generator (see previous review).

sunnysonx commented:
> LGTM. One comment about the API generator (see previous review).

Hello @Elhebert, can you please make the necessary changes if you will have some available time? I don't have access to my pc right now, and I don't want to modify it from mobile without testing 😞  Thank you

Elhebert commented on packages/generators/generators/lib/plops/api.js:
```diff
@@ -45,7 +46,8 @@ module.exports = plop => {
     ],
     actions(answers) {
-      const filePath = answers.isPluginApi && answers.plugin ? 'plugins/{{plugin}}' : 'api/{{id}}';
+      const answerDestination = answers.isPluginApi && answers.plugin ? 'plugin' : 'api';
+      const filePath = getFilePath(answerDestination);
```
```suggestion
      const filePath = answers.isPluginApi && answers.plugin ? 'plugins/{{plugin}}' : 'api/{{id}}';
```

Elhebert reviewed (commented):
Hey @sunnysonx,

Sorry for the late response, I've was out of office for a while. Would you be able to do the requested change now or should I take over it? As this is a PR made from your remote I don't think I can make the change directly on it.

I made a suggestion which will allow you to commit the change from GitHub, thus not needing access to your computer.

sunnysonx commented:
> Hey @sunnysonx,
> 
> Sorry for the late response, I've was out of office for a while. Would you be able to do the requested change now or should I take over it? As this is a PR made from your remote I don't think I can make the change directly on it.
> 
> I made a suggestion which will allow you to commit the change from GitHub, thus not needing access to your computer.

Hello,
Sorry I forgot about this PR. 
Thank you for commit suggestion, I have updated it.

Elhebert commented on packages/generators/generators/lib/plops/api.js:
```diff
@@ -2,6 +2,7 @@
 
 const { join } = require('path');
 const fs = require('fs-extra');
+const getFilePath = require('./utils/get-file-path');
```
The linter is failing because this import is no longer required.

```suggestion
```

Elhebert reviewed (approved):
LGTM, thanks!
